pkg/resources/etcd/statefulset.go

```diff
@@ -314,3 +338,23 @@ func getLauncherImage(data etcdStatefulSetCreatorData) (string, error) {
 	}
 	return data.ImageRegistry(resources.RegistryQuay) + "/kubermatic/etcd-launcher-" + baseTag + ":" + resources.KUBERMATICCOMMIT, nil
 }
+
+func computeReplicas(data etcdStatefulSetCreatorData, set *appsv1.StatefulSet) int {
+	etcdClusterSize := data.Cluster().Spec.EtcdClusterSize
+	if set.Spec.Replicas == nil { // new replicaset
+		return etcdClusterSize
+	}
+	replicas := int(*set.Spec.Replicas)
+	isEtcdHealthy := data.Cluster().Status.ExtendedHealth.Etcd == kubermaticv1.HealthStatusUp
+	// at required size. do nothing
+	if etcdClusterSize == replicas {
+		return replicas
+	}
+	if isEtcdHealthy { // no scaling until we are healthy
+		if etcdClusterSize > replicas {
+			return replicas + 1
+		}
+		return replicas - 1
```

Comment:
Tested with a pre-existing, running cluster with 3 replicas. The statefulset ended up being reduced to 2 replicas, I assume here (.Spec.EtcdClusterSize not set => etcdClusterSize:=0; replicas:=3; isEtcdHealthy:=true => return 2). The etcd launcher then refuses to launch a 2-node cluster because that's smaller than the defaultClusterSize (3), leading to a permanently non-running cluster.

Reply:
No, it's defaulted [here](https://github.com/kubermatic/kubermatic/pull/5571/commits/11068e39b1e90f875b5b62d4b6b96b1fc7d7190e#diff-eea4352ca8918a218965fcc028c01dceR34-R36), which runs as part of the API server. I am not 100% sure if it will default existing clusters as well. But it is a good point, we should be more defensive here.

Reply:
Confirmed that an existing cluster is migrated to the new launcher properly if .spec.etcdClusterSize is set to 3 first.

Reply:
> No, it's defaulted [here](https://github.com/kubermatic/kubermatic/pull/5571/commits/11068e39b1e90f875b5b62d4b6b96b1fc7d7190e#diff-eea4352ca8918a218965fcc028c01dceR34-R36), which runs as part of the API server. I am not 100% sure if it will default existing clusters as well. But it is a good point, we should be more defensive here.

Yeah the defaulting I saw too, but I'm pretty sure that's only [called when new clusters are created](https://github.com/kubermatic/kubermatic/blob/a5e42e57fe7d64bd8a1052b0ec901e4458c72162/pkg/resources/cluster/cluster.go#L59).

Reply:
@multi-io I just pushed a fix for it. Can you please confirm it's working?

Reply:
Confirmed. Existing etcd cluster was migrated without having to manually set .spec.etcdClusterSize or anything else in the cluster resource.

Reply:
@multi-io I think this should work https://github.com/kubermatic/kubermatic/pull/5571/files#diff-60471266513513d81d30f4855c2c01e2R316